locator/token_metadata.cc

```diff
@@ -1261,13 +1259,30 @@ topology::topology(config cfg)
     _pending_locations[utils::fb_utilities::get_broadcast_address()] = std::move(cfg.local_dc_rack);
 }
 
-topology::topology(const topology& other)
-        : _dc_endpoints(other._dc_endpoints)
-        , _dc_racks(other._dc_racks)
-        , _current_locations(other._current_locations)
-        , _pending_locations(other._pending_locations)
-        , _sort_by_proximity(other._sort_by_proximity)
-{
+future<> topology::copy_gently(const topology& other) {
+    co_await clear_gently();
+    _dc_endpoints.reserve(other._dc_endpoints.size());
+    for (const auto& p : other._dc_endpoints) {
+        _dc_endpoints.emplace(p);
+        co_await coroutine::maybe_yield();
+    }
+    _dc_racks.reserve(other._dc_racks.size());
+    for (const auto& [dc, rack_endpoints] : other._dc_racks) {
+        _dc_racks[dc].reserve(rack_endpoints.size());
+        for (const auto& p : rack_endpoints) {
+            _dc_racks[dc].emplace(p);
+            co_await coroutine::maybe_yield();
+        }
+    }
+    for (const auto& p : other._current_locations) {
+        _current_locations.emplace(p);
+        co_await coroutine::maybe_yield();
+    }
+    for (const auto& p : other._pending_locations) {
+        _pending_locations.emplace(p);
+        co_await coroutine::maybe_yield();
+    }
```

Comment:
As is this.

Did we ever see stalls here?

Reply:
We've seen stalls in this are that lead to 052a8d036d8c5ef73cb2ace333def29accde3634
but I admit none were particularly related to topology.

The gist of the fix is 11e74e575cfbe843fd83e17c157694a12802b9bb
I'm okay with dropping the the copy_gently patches for now if you think they are not required.

I wanted to submit them early as I'm working on a series that adds a `node` class that is stored in topology and copying the topology will involve copying also each node object.  Although, as you said we're just talking about possibly a few hundred nodes or maybe low thousand nodes, but since topology changes are rare I think we should lean towards non-interference with on going queries vs. optimizing preemption.

Reply:
Note: for nodes we can use lw_shared_ptr to store them, and const node* to refer to them from topology object methods.

Reply:
That pretty much the direction (using boost::intrusive list with auto_unlink member hooks to list the nodes for different purposes).
But still when cloning the topology as part of the token_metadata for mutating the effective_replication_map, I'd like to copy all node objects rather than use the lw_shared_ptr<node> so that changes in nodes state won't be reflected to previous versions that might be in use.

Reply:
Hmmfmfmfmf. Some node state may need to change (if the node is alive or not).

Reply:
See https://github.com/scylladb/scylladb/pull/11987

Reply:
We can choose: does dead/alive impact effective_replication_map? If it does, it can allow us to skip some computations in the coordinator, but we have to wait for the new computation every time a node flaps. If it does not, then we need to propagate the state. We can even do in between: if the erm live/dead state matches the real state, use it, otherwise do additional filtering.

We can't just kill queries that started with an old replication map. It would introduce errors. The idea is to only change to a "compatible" replication map, and before changing to an incompatibe replication map, wait for anything uses an older, incmpatible map to complete.